api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py

```diff
@@ -121,23 +121,23 @@ def __init__(
         obs_column_names: Sequence[str],
         obs_joinids_chunked: List[npt.NDArray[np.int64]],
         var_joinids: npt.NDArray[np.int64],
+        shuffle_chunk_count: Optional[int] = None,
         shuffle_rng: Optional[Generator] = None,
     ):
         self.obs = obs
         self.X = X
         self.obs_column_names = obs_column_names
-        self.obs_joinids_chunks_iter = self._maybe_local_shuffle_obs_joinids(obs_joinids_chunked, shuffle_rng)
+        if shuffle_chunk_count:
+            assert shuffle_rng is not None
+            splits = len(obs_joinids_chunked) // min(len(obs_joinids_chunked), shuffle_chunk_count)
+            self.obs_joinids_chunks_iter = (
+                shuffle_rng.permutation(np.concatenate(shuffle_chunks))
+                for shuffle_chunks in np.array_split(obs_joinids_chunked, splits)
+            )
+        else:
+            self.obs_joinids_chunks_iter = iter(obs_joinids_chunked)
```

Comment:
should factor this out into a method

Reply:
If I am reading this correctly, if `shuffle_chunk_count = 2`, then this block of code would split the list globally shuffled chunks into 2 lists of chunks where each such list contains say N/2 chunks. It would then concatenate these N/2 chunks in memory and then shuffle this concatenated ndarray in memory.

This might exceed the memory budget and cause an OOM crash.

I think the `splits` calculation in line 132 should be removed, and `shuffle_chunk_count` should be used in place of `splits` in line 135.

So if `shuffle_chunk_count = 2`, it would take 2 chunks at a time and concatenate them and then shuffled the concatenated ndarray which would fit into the memory budget.

Reply:
@atolopko-czi - Please disgard the above comment. I confused the math. Assuming There are `16 chunks` and `shuffle_chunk_count = 2`, you would need `splits = 16/2 = 8` to split the `16 chunks` into 8 partitions of `2 chunks` each.

This makes sense now.

Reply:
yeah, probably worth refactoring this to improve clarity; i'm pushing some variable renames now fwiw